sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/pool/JmsSessionFactory.java

```diff
@@ -0,0 +1,122 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.jms.pool;
+
+import static org.apache.beam.sdk.io.jms.JmsIO.Writer.JMS_IO_PRODUCER_METRIC_NAME;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.JMSException;
+import javax.jms.Session;
+import org.apache.beam.sdk.io.jms.JmsIO;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.commons.pool2.BasePooledObjectFactory;
+import org.apache.commons.pool2.PooledObject;
+import org.apache.commons.pool2.impl.DefaultPooledObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings({
+  "nullness" // TODO(https://github.com/apache/beam/issues/20497)
+})
+public class JmsSessionFactory<T> extends BasePooledObjectFactory<Session> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JmsSessionFactory.class);
+  public static final String CONNECTION_ERRORS_METRIC_NAME = "connectionErrors";
+
+  private final Counter connectionErrors =
+      Metrics.counter(JMS_IO_PRODUCER_METRIC_NAME, CONNECTION_ERRORS_METRIC_NAME);
+
+  private final JmsIO.Write<T> spec;
+  private boolean isConnectionClosed;
+
+  public JmsSessionFactory(JmsIO.Write<T> spec) {
+    this.spec = spec;
+  }
+
+  @Override
+  public Session create() throws JMSException {
+    Connection connection;
+    // reset the connection flag
+    this.isConnectionClosed = false;
+    ConnectionFactory connectionFactory = spec.getConnectionFactory();
+    if (spec.getUsername() != null) {
+      connection = connectionFactory.createConnection(spec.getUsername(), spec.getPassword());
+    } else {
+      connection = connectionFactory.createConnection();
+    }
+    connection.setExceptionListener(
+        exception -> {
+          this.isConnectionClosed = true;
+          this.connectionErrors.inc();
+        });
+    LOG.debug("creating new connection: {}", connection);
+    connection.start();
+    // false means we don't use JMS transaction.
+    return connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+  }
+
+  @Override
+  public boolean validateObject(PooledObject<Session> pooledObject) {
+    return !isConnectionClosed && !callSessionMethod(pooledObject.getObject(), "isClosed", true);
+  }
+
+  @Override
+  public void destroyObject(PooledObject<Session> pooledObject) throws JMSException {
+    Session session = pooledObject.getObject();
+    session.close();
+    Connection connection = callSessionMethod(session, "getConnection", null);
+    if (connection != null) {
+      connection.close();
+    }
+  }
+
+  @Override
+  public PooledObject<Session> wrap(Session session) {
+    return new DefaultPooledObject<>(session);
+  }
+
+  private <U> U callSessionMethod(Session session, String methodName, U defaultValue) {
```

Comment:
Because the Javax API doesn't provide any function for closing the session/connection. But, the two JMS clients (ActiveMQ & qpid) have the same functions getSession & getConnection. The main problem that I found out is that the connection is not getting closed after the session is closed.

>PS: I didn't check other clients

Reply:
I'm using JmsIO with the Solace broker and this reflection usage will most definitely not work with the JCSMP API which Solace uses to implement JMS and probably will break JmsIO usage with Solace altogether... so please consider that checking two clients might not be enough...

Reply:
@ppawel which client are using with the connector `JmsIO` ? We are using Solace as a message broker too but with `Qpid` as a client (The connector Solace is not well maintained). It would be great to add the client that you are using in the test cases so we won't make sure we don't have any regression

Reply:
I am using the official Solace implementation of JMS (sol-jms) but it is giving us a lot of headaches combined with Apache Beam streaming. I might try the Qpid JMS client, but in any case I am looking forward to this PR being integrated as well because currently one of the problems we observe is connection count blowing up under load.